### DISCUSSION TITLE
openzfs 2 and com.delphix:log_spacemap

### DISCUSSION THREAD

**Discussion by okapia** [openzfs/zfs#11920] [Q&A]
In January, I enabled the zfs-testing repository on a RHEL 8 system and upgraded OpenZFS from it to get version 2.0.0. This worked fine for some months and there were no issues. I had created a fresh pool with a special vdev.

I was away for a couple of weeks and a colleague had put kernel updates on the machine. When I got back I had to reinstall the zfs modules, this time getting 2.0.4 from the zfs-testing yum repository but when I come to import the dataset, I get:
```
 state: UNAVAIL
status: The pool can only be accessed in read-only mode on this system. It
	cannot be accessed in read-write mode because it uses the following
	feature(s) not supported on this system:
	com.delphix:log_spacemap (Log metaslab changes on a single spacemap and flush them periodically.)
action: The pool cannot be imported in read-write mode. Import the pool with
	"-o readonly=on", access the pool on a system that supports the
	required feature(s), or recreate the pool from backup.
```
I've uninstalled and reinstalled multiple times to make sure it really is running zfs 2.0.4 and not 0.8.x or something. Did this log_spacemap feature get backed out perhaps? Do I need to pick between a full backup and recreation of the pool or manually building zfs from git to get the feature? Or can I get the old 2.0.0 packages somehow?

**Comment by rincebrain** [comment]
[This](https://openzfs.github.io/openzfs-docs/Basic%20Concepts/Feature%20Flags.html) claims log_spacemap is in 2.0.4. I personally just installed the zfs-testing packages on CentOS 8 (`zpool version` reports zfs-2.0.4-1 and zfs-kmod-2.0.4-1) and created a pool, and confirmed that feature@log_spacemap=active on the new pool.

So are you sure you have zfs 2.0.4? What does `zpool version` say?

**Comment by okapia** [comment]
Ok, that does produce:
```
# zpool version
zfs-2.0.4-1
zfs-kmod-0.8.2-1
```
So despite some repeated dkms commands and attempting to wipe out and reinstall it, 0.8.2 it is still there. Seems `dkms uninstall` didn't work. I've justed tried rm on the `zfs.ko` file and `dkms remove zfs/2.0.4` and now `install` and I'll see if that gets me any further. Thanks

**Reply by rincebrain** [reply]
My suggestion, if dkms install zfs/2.0.4 didn't clobber it, would be to check whether `modinfo zfs` is reporting a different path for the ZFS module than where dkms install is putting 2.0.4, and manually remove it and its associated modules (otherwise you'll probably run into weird dependency issues where you try `modprobe zfs` and it errors out about some symbols in the dependencies not matching.

(An alternate possibility would be if the initrd has the old ZFS modules in it and is configured to load them before exiting the initrd, thus resulting in modules that no number of dkms installs will clobber. A manual mkinitrd would fix that, though.)

**Reply by okapia** [reply]
Thanks. I didn't know about that modinfo trick for locating it. Certainly on initial reinstall attempts, the file I removed was coming back with the same file size as before.

**Comment by okapia** [comment]
I have it working again. Thanks for pointing me in the right direction. I definitely only had kernel sources for 2.0.4 installed but `dkms` was somehow finding an old build of 0.8.2 cached somewhere to install. Further attempts and uninstalling, rebuilding and installing had again failed. Eventually, `dkms unbuild` may have been what got it to do what I wanted.
